src/dataframe/dataframe.jl

```diff
@@ -1317,3 +1317,79 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+function _expandhelper(col, vect, stride, offset)
+    for i in offset:stride:length(vect)
+        push!(col, vect[i])
+    end
+end
+
+function expand(df::AbstractDataFrame, indexcols)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    if !(indexcols isa Symbol || allunique(indexcols))
+        throw(ArgumentError("Elements of $indexcols must be unique"))
+    end
+    sum(nonunique(df, indexcols))>0 && @warn "duplicate rows in input; expand will only return unique combinations"
+    colind = index(df)[indexcols]
+    dummydf = similar(select(df, colind, copycols=false), 0)
+
+    # Create a vect of vectors of unique values in each column
+    uniqueVals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        if any(ismissing, df[!, col])
+            tempcol = vcat(levels(df[col]), missing)
+        else
+            tempcol = levels(df[col])
+        end
+        push!(uniqueVals, tempcol)
+    end
+
+    # Get a long vector of every possible combination
+    collected = collect(Iterators.flatten(Iterators.product(uniqueVals...)))
+
+    for i in 1:length(indexcols)
+        _expandhelper(dummydf[!, i], collected, length(indexcols), i)
+    end
+
+    return dummydf
+end
+
+function _completehelper(expandeddf, colname, fill)
+    col = expandeddf[!, colname]
+    for (i,v) in enumerate(expandeddf[!,names(expandeddf)[end]])
+        v == "left_only" && (col[i] = fill)
+    end
+end
+
+function complete(df::AbstractDataFrame, indexcols; fill=missing, replaceallmissing=false::Bool)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    allunique(indexcols) || throw(ArgumentError("Elements of $indexcols must be unique"))
+    sum(nonunique(df, indexcols))>0 && throw(ArgumentError("duplicate rows in input"))
+    colind = index(df)[indexcols]
+
+    # Expand the input df and left join
+    expanded = expand(df, indexcols)
+    expanded = join(expanded, df; on = indexcols, kind = :left, indicator = :source)
+
+    # Replace missing values with the fill
+    if !ismissing(fill)
+        for n in names(expanded)
```

Comment:
join will first add columns from `expanded` and only next from `df`, so it is enough to iterate columns added, not all columns.

Also note that this will **not** replace `missing` values in the `expaned`-part of the data frame, but I thing this is what we want (i.e. we should replace `missing`s in the joined columns from `df` but not in `expaned` where `missing` is actually a key - if you replaced `missing` with something else - you could - in particular - introduce duplicate rows in the `expanded` subpart of the data frame as the replacement value could be the same as key value actually present in that column)

Reply:
This makes a lot of sense. Decent amount of code optimization happened from this, thank you!